### PR TITLE
Switch CrateMetadata's source_map_import_info from RwLock to Once

### PR DESCRIPTION
**Pull request by spastorino**

### CODE REVIEW

File: src/librustc_metadata/decoder.rs [hunk 1]
```diff
@@ -1290,86 +1290,69 @@ impl<'a, 'tcx> CrateMetadata {
...
-                   local_version.name, start_pos, end_pos,
-                   local_version.start_pos, local_version.end_pos);
-
-            cstore::ImportedSourceFile {
-                original_start_pos: start_pos,
-                original_end_pos: end_pos,
-                translated_source_file: local_version,
-            }
-        }).collect();
+    ) -> &Vec<cstore::ImportedSourceFile> {
```
Comment (Mark-Simulacrum) [inline src/librustc_metadata/decoder.rs]: I usually prefer to see `&[...]` in cases like this since you almost certainly don't care about the capacity (it also avoids spurious `.clone()`.

File: src/librustc_metadata/decoder.rs [hunk 2]
```diff
@@ -1290,86 +1290,69 @@ impl<'a, 'tcx> CrateMetadata {
...
+                cstore::ImportedSourceFile {
+                    original_start_pos: start_pos,
+                    original_end_pos: end_pos,
+                    translated_source_file: local_version,
+                }
+            }).collect()
+        });
 
-        // This shouldn't borrow twice, but there is no way to downgrade RefMut to Ref.
         self.source_map_import_info.borrow()
```
Comment (Mark-Simulacrum) [inline src/librustc_metadata/decoder.rs]: note: we should make `init_locking` and friends return &T, either here in a separate commit or a separate PR

**Review by Mark-Simulacrum — APPROVED** [review]

**Comment by Mark-Simulacrum** [comment]
r=me modulo comments

File: src/librustc_data_structures/sync.rs [hunk 3]
```diff
@@ -497,13 +497,13 @@ impl<T> Once<T> {
...
         let mut lock = self.0.lock();
-        if lock.is_some() {
-            return false;
+        if lock.is_none() {
+            *lock = Some(f());
         }
-        *lock = Some(f());
-        true
+
+        self.borrow()
```
Comment (Mark-Simulacrum) [inline src/librustc_data_structures/sync.rs]: hm, I suspect this won't work? i.e., the lock appears to still be held -- we can probably move the rest of the body into `{ ... }` or so

Comment (spastorino) [reply src/librustc_data_structures/sync.rs]: Ouch, fixing it.

File: src/librustc_data_structures/sync.rs [hunk 4]
```diff
@@ -497,13 +497,15 @@ impl<T> Once<T> {
     /// If the value was already initialized the closure is not called and `false` is returned,
     /// otherwise if the value from the closure initializes the inner value, `true` is returned
     #[inline]
```
Comment (therealprof) [inline src/librustc_data_structures/sync.rs]: Shouldn't the comment be changed to reflect the change that instead of returning a `bool` we now return a `&T`?

Comment (spastorino) [reply src/librustc_data_structures/sync.rs]: Definitely :), going to do that now.
